File: SpreadsheetLightDataGridViewExport/Classes/ExcelOperations.cs
using System;
using System.Collections.Generic;
using System.Data;
using System.IO;
using System.Linq;
using SpreadsheetLight;
using static System.DateTime;

namespace SpreadsheetLightDataGridViewExport.Classes
{

    public class ExcelOperations
    {
        /// <summary>
        /// File name to write DataTable to
        /// </summary>
        public static string FileName { get; set; }
        public static string SheetName = $"Orders_{Now.Month}_{Now.Year}";

        public static string Find(string fileName ,string sheetName, string search)
        {
            using (var document = new SLDocument(fileName,sheetName))
            {
                var stats = document.GetWorksheetStatistics();
                var test = stats.EndRowIndex;
                for (int index = 1; index < stats.EndRowIndex +1; index++)
                {
                    if (document.GetCellValueAsString(index, 1).EqualsIgnoreCase(search))
                    {
                        var result = document.GetCellValueAsString(index, 2);
                        return result;
                    }
                }
            }

            return null;
            
        }
        public static List<int> FindDuplicates(string fileName, string sheetName, string search)
        {
            List<int> indicesList = new List<int>();
            
            using (var document = new SLDocument(fileName, sheetName))
            {
                var stats = document.GetWorksheetStatistics();
                var test = stats.EndRowIndex;
                for (int index = 1; index < stats.EndRowIndex + 1; index++)
                {
                    if (document.GetCellValueAsString(index, 1).EqualsIgnoreCase(search))
                    {
                        indicesList.Add(index);

                    }
                }
            }

            return indicesList;
        }

        /// <summary>
        /// Here we use create a new file although with minor modifications,
        /// an existing file/sheet can be used and append to the existing sheet
        /// or with an existing file add a new sheet.
        /// </summary>
        /// <param name="table">DataTable from DataGridView</param>
        /// <remarks>
        /// Common error is having <see cref="FileName"/> open when performing operations
        /// below
        /// </remarks>
        public static (bool success, Exception exception) Export(DataTable table)
        {
            try
            {
                if (File.Exists(FileName))
                {
                    File.Delete(FileName);
                }

                using (var document = new SLDocument())
                {
                    // style for date columns
                    SLStyle dateStyle = document.CreateStyle();
                    dateStyle.FormatCode = "mm-dd-yyyy";

                    /*
                     * import DataTable without column names,
                     * change last param to true for column names
                     */
                    document.ImportDataTable(1, SLConvert.ToColumnIndex("A"), table, true);

                    // apply date style to DateTime columns
                    document.SetColumnStyle(4, dateStyle);
                    document.SetColumnStyle(5, dateStyle);
                    document.SetColumnStyle(6, dateStyle);

                    // Auto format date columns
                    document.AutoFitColumn(4);
                    document.AutoFitColumn(5);
                    document.AutoFitColumn(6);

                    // rename default sheet, sheet1
                    document.RenameWorksheet(
                        SLDocument.DefaultFirstSheetName,
                        SheetName);

                    // save to file
                    document.SaveAs(FileName);

                    return (true, null);
                }
            }
            catch (Exception exception)
            {
                return (false, exception);
            }
        }

        /// <summary>
        /// https://stackoverflow.com/questions/70378875/best-way-to-export-large-datatable-into-excel-in-c/70379151#70379151
        /// </summary>
        /// <param name="table"></param>
        /// <returns></returns>
        public static (bool success, Exception exception) ExportDataTable(DataTable table)
        {
            try
            {

                using (var document = new SLDocument())
                {
                    document.ImportDataTable(1, SLConvert.ToColumnIndex("A"), table, true);
                    document.SaveAs(FileName);

                    return (true, null);
                }
            }
            catch (Exception exception)
            {
                return (false, exception);
            }
        }
    }

    public class SpreadSheetLightDemos
    {
        public static (bool success, Exception exception) Export(DataTable table, string fileName, string sheetName)
        {
            try
            {
                using (var document = new SLDocument())
                {
                    document.ImportDataTable(1, SLConvert.ToColumnIndex("A"), table, true);
                    document.RenameWorksheet(SLDocument.DefaultFirstSheetName, sheetName);
                    document.SaveAs(fileName);

                    return (true, null);
                }
            }
            catch (Exception exception)
            {
                return (false, exception);
            }
        }

        /// <summary>
        /// Open file to specific sheet
        /// </summary>
        /// <param name="fileName">Existing Excel file name</param>
        /// <param name="sheetName">Existing Sheet name in file</param>
        public static void Work(string fileName, string sheetName)
        {
            using (var document = new SLDocument(fileName, sheetName))
            {
                // some code do write to the cells of selected sheet
            }
        }
    }
}

File: SpreadSheetLightVb1/ExcelClasses/ExcelOperations.vb
Imports System.IO
Imports SpreadsheetLight

Public Class ExcelOperations

    ''' <summary>
    ''' Export a DataTable to Excel
    ''' </summary>
    ''' <param name="pFileName">path and file name to save too, path is optional</param>
    ''' <param name="pSheetName">Name of sheet</param>
    ''' <param name="pDataTable">Populated DataTable</param>
    ''' <param name="pColumnHeaders">True for column names as first row, false no column names</param>
    Public Shared Sub SimpleExportRaw(pFileName As String, pSheetName As String, pDataTable As DataTable, pColumnHeaders As Boolean)

        Using doc As New SLDocument()

            doc.SelectWorksheet(pSheetName)

            doc.ImportDataTable(1, SLConvert.ToColumnIndex("A"), pDataTable, pColumnHeaders)

            SLConvert.ToColumnName(1)

            Dim style As New SLStyle With {.FormatCode = "MM/dd//yyyy"}

            ' Format modified date column
            doc.SetColumnStyle(9, style)

            style.Font.Bold = True
            doc.SetRowStyle(1, 1, style)

            Dim stats = doc.GetWorksheetStatistics()

            ' auto fit all columns
            doc.AutoFitColumn(1, stats.EndColumnIndex)

            ' original default name is Sheet1, let's change it to the name in pSheetName
            doc.RenameWorksheet(SLDocument.DefaultFirstSheetName, pSheetName)

            doc.SaveAs(pFileName)

        End Using

    End Sub
    Public Delegate Sub OnErrorDelegate(exception As Exception)
    Public Shared Event OnErrorEvent As OnErrorDelegate
    ''' <summary>
    ''' DemoExport.xlsx
    ''' Info_Table
    ''' </summary>
    ''' <param name="pFileName"></param>
    ''' <param name="pSheetName"></param>
    ''' <param name="pDataTable"></param>
    ''' <param name="pColumnHeaders"></param>
    ''' <param name="pStartRow"></param>
    Public Shared Function Export(pFileName As String, pSheetName As String, pDataTable As DataTable, pColumnHeaders As Boolean, Optional pStartRow As Integer = 3) As Boolean

        Try

            CopyFile(pFileName)

            Using doc As New SLDocument(pFileName)

                doc.SelectWorksheet(pSheetName)
                doc.ImportDataTable(pStartRow, SLConvert.ToColumnIndex("A"), pDataTable, pColumnHeaders)

                Dim style As New SLStyle With {.FormatCode = "MM/dd//yyyy"}


                If pDataTable.Columns.Contains("ModifiedDate") Then
                    doc.SetColumnStyle(pDataTable.Columns("ModifiedDate").Ordinal + 1, style)
                End If

                doc.Save()

                Return True

            End Using

        Catch ex As Exception
            RaiseEvent OnErrorEvent(ex)
            Return False
        End Try

    End Function
    Public Shared Sub Export(
         pFileName As String, 
         pSheetName As String, 
         pDataSet As DataSet, 
         pTableName As String, 
         pColumnHeaders As Boolean)

        Using doc As New SLDocument()

            doc.SelectWorksheet(pSheetName)
            doc.ImportDataTable(1, SLConvert.ToColumnIndex("A"), pDataSet.Tables(pTableName), pColumnHeaders)
            doc.RenameWorksheet(SLDocument.DefaultFirstSheetName, pSheetName)
            doc.SaveAs(pFileName)

        End Using

    End Sub
    Public Shared Sub ExportTable(pFileName As String, pSheetName As String, pDataTable As DataTable, pColumnHeaders As Boolean)

        Using doc As New SLDocument()
            doc.SelectWorksheet(pSheetName)
            doc.ImportDataTable(1, SLConvert.ToColumnIndex("A"), pDataTable, pColumnHeaders)
            doc.RenameWorksheet(SLDocument.DefaultFirstSheetName, pSheetName)
            doc.SaveAs(pFileName)
        End Using

    End Sub
    Public Shared Sub InsertTable()


    End Sub
    Private Shared Sub CopyFile(pFileName As String)
        Dim originalFile = Path.Combine(AppDomain.CurrentDomain.BaseDirectory, "ExcelFiles", pFileName)
        Dim targetFile = Path.Combine(AppDomain.CurrentDomain.BaseDirectory, pFileName)

        File.Copy(originalFile, targetFile, True)
    End Sub

    ''' <summary>
    ''' Mocked code sample to load a list
    ''' The try-catch is mainly here to ensure we don't happen
    ''' to run this code with the file open from a prior run
    '''
    ''' Option Strict On
    ''' Option Infer On
    ''' 
    ''' </summary>
    ''' <returns></returns>
    Public Shared Function CreateAndPopulate() As Boolean

        Dim success = True
        Try
            Using doc As New SLDocument()

                Dim style As New SLStyle

                style.Font.Bold = True
                doc.SetRowStyle(1, 1, style)

                doc.SetCellValue("A1", "Last name")
                doc.SetCellValue("B1", "First name")

                Dim people = PeopleData.List()

                Dim rowIndex = 0

                For Each person In people
                    doc.SetCellValue($"A{rowIndex + 2}", people(rowIndex).LastName)
                    doc.SetCellValue($"B{rowIndex + 2}", people(rowIndex).FirstName)
                    rowIndex += 1
                Next


                Dim stats = doc.GetWorksheetStatistics()

                doc.AutoFitColumn(1, stats.EndColumnIndex)
                doc.RenameWorksheet(SLDocument.DefaultFirstSheetName, "People")


                doc.SaveAs("PeopleData.xlsx")

            End Using
        Catch ex As Exception
            Console.WriteLine($"{ex.Message}")
            success = False
        End Try

        Return success

    End Function

    ''' <summary>
    ''' Writes three values to the default worksheet
    ''' </summary>
    ''' <param name="fileName"></param>
    ''' <param name="value1"></param>
    ''' <param name="value2"></param>
    ''' <param name="value3"></param>
    Public Shared Sub SimpleWrite(fileName As String, value1 As String, value2 As String, value3 As String)

        If Not File.Exists(fileName) Then
            Throw New FileNotFoundException($"Dude failed to find {fileName}")
        End If

        If Not String.IsNullOrWhiteSpace(value1) AndAlso Not String.IsNullOrWhiteSpace(value2) AndAlso Not String.IsNullOrWhiteSpace(value3) Then

            Try

                Using doc As New SLDocument(fileName)

                    doc.SetCellValue($"{SLConvert.ToColumnName(1)}2", value1)
                    doc.SetCellValue($"{SLConvert.ToColumnName(1)}3", value2)
                    doc.SetCellValue($"{SLConvert.ToColumnName(1)}4", value3)

                    doc.Save()

                End Using
            Catch ex As Exception
                Console.WriteLine(ex.Message)
            End Try

        Else
            Console.WriteLine("Missing one or more values")
        End If

    End Sub
    Public Shared Sub SimpleWrite(fileName As String, workSheetName As String, value1 As String, value2 As String, value3 As String)

        If Not File.Exists(fileName) Then
            Throw New FileNotFoundException($"Dude failed to find {fileName}")
        End If

        If Not String.IsNullOrWhiteSpace(value1) AndAlso Not String.IsNullOrWhiteSpace(value2) AndAlso Not String.IsNullOrWhiteSpace(value3) Then

            Try

                Using doc As New SLDocument(fileName)

                    If SheetExists(doc, workSheetName) Then
                        doc.SelectWorksheet(workSheetName)
                    End If


                    doc.SetCellValue($"{SLConvert.ToColumnName(1)}2", value1)
                    doc.SetCellValue($"{SLConvert.ToColumnName(1)}3", value2)
                    doc.SetCellValue($"{SLConvert.ToColumnName(1)}4", value3)

                    doc.Save()

                End Using
            Catch ex As Exception
                Console.WriteLine(ex.Message)
            End Try

        Else
            Console.WriteLine("Missing one or more values")
        End If

    End Sub
    Public Shared Function SheetExists(doc As SLDocument, pSheetName As String) As Boolean
        Return doc.GetSheetNames(False).Any(Function(sheetName) sheetName.ToLower() = pSheetName.ToLower())
    End Function
End Class